Vectorize add_phylopic functions and add geom_phylopic

willgearty opened the pull request:
This PR adds a new `geom_phylopic()` which acts like `geom_point()`, but places PhyloPic silhouettes instead. The silhouettes can be specified with "name", "uuid", or actual image objects (svg or png).

Note that the ggimage package also has a function called "geom_phylopic", but this function has several advantages:

- **Supports vector images** (and this is what is downloaded if using "name" or "uuid" values)
- Allows the use of already downloaded silhouettes
- Height ("size") is specified in y-axis units (ggimage uses y-axis percentage)
- Allows for horizontal/vertical flipping  (ggimage supports rotation)

I suppose we could call this something else, but I don't think there will be too much usage overlap.

In addition to the new function, I've also vectorized both `add_phylopic` (which uses `ggplot2::annotate` and the new `GeomPhylopic`) and `add_phylopic_base` (which now uses `mapply` to allow any argument to be vectorized).

Fixes https://github.com/palaeoverse-community/rphylopic/issues/25 and Fixes https://github.com/palaeoverse-community/rphylopic/issues/42.

willgearty commented:
Now supports polar format (including the ggtree "circular" layout) as requested by @bschilder.

bschilder commented:
Amazing! Thanks @willgearty !

Regarding the rectangle issue, this comes from the fact that some images have a filled background and others have a  transparent background. 

While the ultimate solution would be to have consistency with the phylpic database, @xiangpin came up with a fix using `magick` to remove the background on the fly:
https://github.com/GuangchuangYu/ggimage/pull/46

willgearty commented:
Thanks for pointing me to that PR @bschilder. Unfortunately, the default functionality for rphylopic is to handle vector images (svgs), so `magick::image_convert()` won't work in this case. However, I have added the option to use the "original" color for now.

LewisAJones reviewed (commented):
Well this has really come along! Great job @willgearty, I really like the updates and everything is working as expected. I've been through the code and left a few general comments. Everything is minor though and mainly suggestions for improving documentation and addressing some lint issues. Once everything is merged, we should ask Miranta to update the cheatsheet for us.

willgearty commented on R/add_phylopic.r:
```diff
@@ -79,71 +75,34 @@ add_phylopic <- function(img = NULL, name = NULL, uuid = NULL,
-      pars$alpha <- alpha
-      pars
+  # Make all variables the same length
+  x_len <- length(x)
```
Yeah, I had thought about this, but it does seem like quite an odd use case and there is a very easy workaround, as you mention. This is the only real difference between this and add_phylopic_base.

willgearty commented on R/geom_phylopic.R:
```diff
@@ -0,0 +1,214 @@
+#'
+#' - **x**
+#' - **y**
+#' - img/uuid/name
```
Bold is the ggplot2 standard. Do you still want them underlined?

LewisAJones replied:
I think it would be clearer and easier to read for the user

willgearty commented on R/geom_phylopic.R:
```diff
@@ -0,0 +1,214 @@
+                          na.rm = FALSE,
+                          show.legend = FALSE,
+                          inherit.aes = TRUE) {
+  layer(
```
The ggplot2 internals should handle most of the error handling here, and there is additional error_handling in `GeomPhylopic`.

willgearty commented:
Thanks for the thorough review @LewisAJones! I've fixed the handling of alpha values (now we just use `recolor_phylopic` instead of doing some funky/buggy stuff within `grid.picture`), fixed the handling of white backgrounds (https://github.com/keesey/phylopic/issues/14) which reduced the need for extra details about the color argument, and fixed a new bug related to highly nested svg objects (I think that may be my first recursive function that I've written in R).

Let me know what you think!

LewisAJones reviewed (approved):
Hey @willgearty, this looks great! I really like your efficient solution to removing the white background for images. I don't really have anything to add which would need implementing so I think this is ready to merge. Just a reminder to update NEWS with the additional functionality for removing the background (this is relevant for `add_phylopic` and `add_phylopic_base`).
